pubsub/index.js

```diff
@@ -6,239 +6,172 @@
-      clearMessages(payload, message => message.ack())
-      console.log(timestamp(), 'Success!', endpoint, payload.length)
-    } catch (error) {
-      console.log(timestamp(), endpoint, error.stack)
-      clearMessages(payload, message => message.nack(), true)
-    }
-  }, MAX_EVENTS_PER_BATCH)
-
-  cargo.concurrency = WORKER_COUNT
-
-  return cargo
-}
-
-function processMessage (cargo, message) {
-  const { httpapi, identify } = parseMessage(message)
 
-  if (message.publishTime < Date.now() - WARNING_THRESHOLD) {
-    console.log(timestamp(), 'Warning: Old message', { httpapi, identify })
-  }
-
-  if (httpapi) {
-    MESSAGES.set(httpapi.insert_id, { message, payloadCount: identify ? 2 : 1 })
-    cargo.httpapi.push(httpapi)
-  }
-
-  if (identify) {
-    cargo.identify.push(identify)
-  }
-}
-
-function parseMessage (message) {
-  let { jsonPayload: event } = JSON.parse(Buffer.from(message.data, 'base64').toString())
+  const cargo = async.cargo(async (payload) => {
+    let identify_events = []
+    let http_events = payload.map(message => {
+      let event = JSON.parse(Buffer.from(message.data, 'base64').toString()).jsonPayload
+
+      if (event.Fields) {
+        event = event.Fields
+    
+        if (event.op && event.data) {
+          event = JSON.parse(event.data)
+        } else {
+          if (is.nonEmptyString(event.event_properties)) {
+            event.event_properties = JSON.parse(event.event_properties)
+          }
+    
+          if (is.nonEmptyString(event.user_properties)) {
+            event.user_properties = JSON.parse(event.user_properties)
+          }
+        }
+      }
 
-  if (event.Fields) {
-    event = event.Fields
+      if (! isEventOk(event)) {
+        logger.warn({ type: 'event.malformed', event }, 'Dropping malformed event')
+        message.ack()
+        return null
+      }
 
-    if (event.op && event.data) {
-      event = JSON.parse(event.data)
-    } else {
-      if (is.nonEmptyString(event.event_properties)) {
-        event.event_properties = JSON.parse(event.event_properties)
+      if (event.user_id) {
+        event.user_id = hash(event.user_id)
+      }
+    
+      event.insert_id = hash(event.user_id, event.device_id, event.session_id, event.event_type, event.time)
+      
+      if (IDENTIFY_VERBS.some(verb => is.assigned(event.user_properties[verb]))) {
+        identify_events.push({
+          device_id: event.device_id,
+          user_id: event.user_id,
+          user_properties: splitIdentifyPayload(event.user_properties),
+        })
       }
 
-      if (is.nonEmptyString(event.user_properties)) {
-        event.user_properties = JSON.parse(event.user_properties)
+      return event
+    }).filter(message => !!message)
+
+    // Filter out exact duplicate identify events to help prevent rate limiting
+    identify_events = identify_events.filter((outer_event, index, self) => {
+      return index === self.findIndex((inner_event) => {
```

Comment:
So we are trying to filter out duplicate entries, but I feel like what this does is just return identify_events unfiltered. For each item in the outer loop, it's going to find a match in the inner loop, and that match is going to be itself. 

On second read, if there is a second entry, it's going to find the first entry first, and the returned index is going to not be equal, and that item will be filtered out. 👍 



I wonder if this would read more clearly if we used `reduce`, or even went straight up imperative:

### reduce:
```js
identify_events = identify_events.reduce((accumulator, outer_event) => {
  if (! accumulator.find(inner_event => same(outer_event, inner_event)) {
    accumulator.push(outer_event);
  });
  return accumulator;
}, []);
```
### imperative
```js
const accumulator = [];
for (let outer_event of identify_events) {
  if (! accumulator.find(inner_event => same(outer_event, inner_event)) {
    accumulator.push(outer_event);
  }
}
```
Maybe not.

I also wonder if 2*O(n^2) is going to be problematic since we process so many events.